### PR TITLE
Extremely WIP symmetric encryption

### PR DESCRIPTION
**Pull request by DavidGregory084**
Raising this mostly for discussion - are you interested in fleshing this library out with symmetric crypto algorithms?

~~At the moment this extremely doesn't work because I don't know how to read test resources files in Scala.js~~ Fixed in https://github.com/typelevel/bobcats/pull/115/commits/4e6cedf6144d791ddebf795b9f1360745b47753a

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -94,8 +94,10 @@ lazy val core = crossProject(JSPlatform, JVMPlatform)
     libraryDependencies ++= Seq(
       "org.typelevel" %%% "cats-core" % catsVersion,
       "org.typelevel" %%% "cats-effect-kernel" % catsEffectVersion,
+      "org.typelevel" %%% "cats-effect-std" % catsEffectVersion,
```
Comment (DavidGregory084) [inline build.sbt:97]: Needed for `Random.javaSecuritySecureRandom`

File: core/shared/src/test/scala/bobcats/AESTestVectorParser.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  private val encryptSection = section("ENCRYPT", encryptEntry)
+
+  private val decryptSection = section("DECRYPT", decryptEntry)
+
+  private val parser =
+    header *> (encryptSection ~ decryptSection).surroundedBy(whitespaces).map(TestVectors.tupled)
+
+  def parse(input: String) =
+    parser.parseAll(input)
+}
```
Comment (DavidGregory084) [inline core/shared/src/test/scala/bobcats/AESTestVectorParser.scala:72]: I wrote this when I saw the format for the initial set of test vector files. However, it seems like the format of those files is absolutely all over the place - the GCM test vector files are totally different.

File: build.sbt (hunk 1 above)
Comment (armanbilge) [reply build.sbt:97]: See also:
- https://github.com/typelevel/bobcats/issues/82

**Comment by armanbilge** [comment]
> are you interested in fleshing this library out with symmetric crypto algorithms?

So far development has been driven based on what's needed for cross-building downstreams, but the long term vision has definitely been to make this _the_ crypto lib for Typelevel :) I'm so glad you've taken an interest!

My only caveat is that I'm very strapped for maintainer time, sorry 😂  please request to join the team!

https://github.com/orgs/typelevel/teams/bobcats/members

See also David's vision for the lib in https://github.com/http4s/http4s/issues/5044#issuecomment-896885041, which I haven't really made a proper attempt of tbh.

> At the moment this extremely doesn't work because I don't know how to read test resources files in Scala.js

Yeah, if we want to test in browsers then resource files are going to be a problem. Can we write a source generator to make test cases from those resources?

**Comment by armanbilge** [comment]
@isomarcte that's great news!! 😁

**Comment by DavidGregory084** [comment]
> Yeah, if we want to test in browsers then resource files are going to be a problem. Can we write a source generator to make test cases from those resources?

@armanbilge I have done this in https://github.com/typelevel/bobcats/pull/115/commits/4e6cedf6144d791ddebf795b9f1360745b47753a

File: project/AESCBCTestVectorGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+          )
+          """
+        }
+      }
+      .map { testVectorDeclarations =>
+        val testVectorDeclarationNames =
+          testVectorDeclarations.collect { case Defn.Val(_, List(Pat.Var(name)), _, _) => name }
+
+        val allTestVectors =
+          q"lazy val allTestVectors = List(..${testVectorDeclarationNames})"
```
Comment (DavidGregory084) [inline project/AESCBCTestVectorGenerator.scala:65]: Without `lazy` we create a static initializer that is too large for the JVM 😛

File: build.sbt [hunk 4]
```diff
@@ -94,13 +95,38 @@ lazy val core = crossProject(JSPlatform, JVMPlatform)
...
+      val generator = FileFunction.cached(
+        streams.value.cacheDirectory / "cbcTestVectors",
+        inStyle = FileInfo.hash,
+        outStyle = FileInfo.exists) { inputs =>
+        AESCBCTestVectorGenerator.generate(
+          inputs,
+          (Test / sourceManaged).value,
+          scalafmtConfigFile
+        )
+      }
```
Comment (DavidGregory084) [inline build.sbt:125]: I think this could be improved by tracking the sources of the generator itself

Comment (armanbilge) [reply build.sbt:125]: Nice! Although this setup seems different than what I've done in the past. Could we use the `fileInputs` key here instead?

https://github.com/armanbilge/gcp4s/blob/6263842d44b187f4e587d5e351cb6e47f164c772/project/DiscoveryPlugin.scala#L31

Comment (DavidGregory084) [reply build.sbt:125]: Hmm there really is no end of ways to do the same thing in sbt is there? I haven't seen this `fileInputs` thing before, will take a look.

**Comment by yilinwei** [comment]
I'd like to resurrect this PR. @DavidGregory084 Are you OK with me forking it and continuing?

**Comment by DavidGregory084** [comment]
Yeah of course @yilinwei, please go ahead 🚀

**Pull request closed without merging**